pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,348 @@
+//
+// The library will log in two different ways once to the standard cockroach log
+// and once to the SQL trace on the leader.
+// TODO(baptist): Look at logging traces on followers as well and sending back
+// to the leader in the MsgAppResp. It would need to be best effort, but might
+// still be useful.
+type RaftTracer struct {
+	// m is a map of all the currently traced entries for this replica. The
+	// aggregate size of the map is equal to or less than numRegistered. We add
+	// to numRegistered before we update m, and delete from m before we remove
+	// from numRegistered to keep this invariant.
+	m syncutil.Map[kvpb.RaftIndex, traceValue]
+	// numRegistered is the number of currently registered traces for this
+	// store, not this replica. The number of registered will always be less
+	// than the MaxConcurrentRaftTraces setting. If the setting is lowered, we
+	// flush all traces on all replicas.
+	numRegistered *atomic.Int64
+	// This is the ambient context for the replica and is used for remote
+	// traces. It contains the replica/range information. On each trace we
+	// additionally append the unique trace/span IDs.
+	ctx context.Context
+	st  *cluster.Settings
+}
+
+// NewRaftTracer creates a new RaftTracer with the given ambient context for the
+// replica.
+func NewRaftTracer(
+	ctx context.Context, st *cluster.Settings, numRegistered *atomic.Int64,
+) *RaftTracer {
+	return &RaftTracer{ctx: ctx, st: st, numRegistered: numRegistered}
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages. 10 is a
+// reasonable default that balances usefulness with performance impact. It isn't
+// expected that this limit will normally be hit.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemVisible,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces",
+	0,
+	settings.NonNegativeInt,
+)
+
+// maybeRegister checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The invariant is
+// that numRegistered <= numAllowed.  This method will return true if we can
+// keep the invariant and added one to the number registered, otherwise it will
+// return false.
+func (r *RaftTracer) maybeRegister() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegistered := r.numRegistered.Load()
+
+	// The maximum number of traces has been reached. We don't register this
+	// trace and return false.
+	if numRegistered == numAllowed {
+		return false
+	}
+
+	// This can only happen if numAllowed has changed. If this happens flush all
+	// our current traces and don't register this request.
+	if numAllowed == 0 {
```

Comment:
Slightly worried about the overhead here. We would be essentially calling `FlushAll` on every message. I realize this `== 0` check is what I previously suggested, sorry.

If we roll back to `if numRegistered > numAllowed { flush() }`, this would also be not ideal: if one or two ranges are not clearing themselves (e.g. it's quiesced or something, and observes no messages), then this check will also incur a flush on every message on all ranges.

Can we find a way to make the 0 case nearly free (not more than one or two atomic loads)? I'm pretty sure we need a variable local to this `RaftTracer` to achieve something like this.

Reply:
I cleaned this up again. Now we have the two variables and check the "local" one first. We will only load the global one if we are likely to register. Also I think the overhead of reading a shared atomic should be ~ the cost of reading the setting. I ran some tests and I don't see it in perf traces.